### PR TITLE
access: fix foreign key declaration

### LINKED ISSUE #3099: PostgreSQL: access module assignes -1 to a foreign key
On `database create` with PostgreSQL I have a problem to resolve:

```python
INFO in dbquery_postgresql [/home/vagrant/.virtualenvs/invenio2/src/invenio/invenio/legacy/dbquery_postgresql.py:114]:
INSERT INTO "accROLE_accACTION_accARGUMENT" ("id_accROLE",
                "id_accACTION" , "id_accARGUMENT", argumentlistid)
                VALUES (%s, %s, -1, -1) 
--------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/.virtualenvs/invenio2/bin/webaccessadmin", line 9, in <module>
    load_entry_point('invenio==2.1.0.dev20150305', 'console_scripts', 'webaccessadmin')()
  File "/home/vagrant/.virtualenvs/invenio2/src/invenio/invenio/base/helpers.py", line 50, in decorated_func
    result = f(*args, **kwargs)
  File "/home/vagrant/.virtualenvs/invenio2/src/invenio/invenio/modules/access/scripts/webaccessadmin.py", line 111, in main
    DEF_DEMO_ROLES, DEF_DEMO_AUTHS)
  File "/home/vagrant/.virtualenvs/invenio2/src/invenio/invenio/modules/access/control.py", line 1913, in acc_add_default_settings
    **args)
  File "/home/vagrant/.virtualenvs/invenio2/src/invenio/invenio/modules/access/control.py", line 541, in acc_add_authorization
    VALUES (%s, %s, -1, -1) """, (id_role, id_action))
  File "/home/vagrant/.virtualenvs/invenio2/src/invenio/invenio/legacy/dbquery_postgresql.py", line 118, in run_sql
    cur = engine.execute(sql, (param, ))
  File "/home/vagrant/.virtualenvs/invenio2/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1986, in execute
    return connection.execute(statement, *multiparams, **params)
  File "/home/vagrant/.virtualenvs/invenio2/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 906, in execute
    return self._execute_text(object, multiparams, params)
  File "/home/vagrant/.virtualenvs/invenio2/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1054, in _execute_text
    statement, parameters
  File "/home/vagrant/.virtualenvs/invenio2/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1146, in _execute_context
    context)
  File "/home/vagrant/.virtualenvs/invenio2/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1339, in _handle_dbapi_exception
    exc_info
  File "/home/vagrant/.virtualenvs/invenio2/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 199, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File "/home/vagrant/.virtualenvs/invenio2/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1139, in _execute_context
    context)
  File "/home/vagrant/.virtualenvs/invenio2/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (psycopg2.IntegrityError) insert or update on table "accROLE_accACTION_accARGUMENT" violates foreign key constraint "accROLE_accACTION_accARGUMENT_id_accARGUMENT_fkey"
DETAIL:  Key (id_accARGUMENT)=(-1) is not present in table "accARGUMENT".
 [SQL: 'INSERT INTO "accROLE_accACTION_accARGUMENT" ("id_accROLE",\n                "id_accACTION" , "id_accARGUMENT", argumentlistid)\n                VALUES (%s, %s, -1, -1) '] [parameters: (6, 51)]
ERROR: failed execution of /home/vagrant/.virtualenvs/invenio2/bin/webaccessadmin -u admin -c -a -D
```

Access module try to put -1 on column `id_accARGUMENT` of the table `accROLE_accACTION_accARGUMENT`.
But it's a column that contains a Foreign Key!
So, PostgreSQL refuses to execute the query... T_T

I have 2 possibility to fix the problem:

1. Remove the Foreign Key from column `id_accARGUMENT` (easy way)
2. Find and substitute `-1` with None in `invenio/modules/access/control.py` (it's full of `run_sql`)

Which it's the best way? 

p.s. considering that I think `access` module requires a heavy refactoring.

cc @jirikuncar

### PR DESCRIPTION
**Pull request by hachreak**
* Fixes sql queries to improve standard compliance and compatibility:
  "RLIKE", quoting. (addresses #3214)

* Fixes problem of assignement of -1 to a foreign key, removing foreign
  key and shaping relationship to exclude -1 value. (closes #3099)

* PEP8/257 code style improvements.

Signed-off-by: Leonardo Rossi <leonardo.r@cern.ch>

### CODE REVIEW

**Comment by hachreak** [comment]
ping @jirikuncar 
Can we start to review the PR? :)

**Comment by hachreak** [comment]
with @jacquerie we are trying to resolve some issues with `inveniomanage demosite populate --yes-i-know`

File: .travis.yml [hunk 1]
```diff
@@ -65,16 +72,17 @@ install:
   - "sudo su -c \"npm install --silent -g bower less clean-css uglify-js requirejs\""
   # All the step below this points are solely for test purposes, don't use them
   # to setup your invenio installation. Please do RTFM instead (INSTALL.rst).
-  - "./scripts/setup_devmode.sh"
+  - "./scripts/setup_devmode.sh $DB"
 
 before_script:
+  - "psql -U postgres -c 'CREATE EXTENSION pgcrypto;'"
```
Comment (jirikuncar) [inline .travis.yml]: we should probably do it only when `[ "$DB" = "postgresql"]`

File: invenio/modules/workflows/testsuite/test_workflows.py [hunk 2]
```diff
@@ -106,7 +110,9 @@ def setUp(self):
     def tearDown(self):
         """ Clean up created objects."""
         from invenio.modules.workflows.models import Workflow
-        Workflow.get(Workflow.module_name == "unit_tests").delete()
+        self.delete_objects(
+            Workflow.get(Workflow.module_name == "unit_tests").all())
+        # Workflow.get(Workflow.module_name == "unit_tests").delete()
```
Comment (jirikuncar) [inline invenio/modules/workflows/testsuite/test_workflows.py]: you can remove this line.

File: invenio/modules/sequencegenerator/texkey.py [hunk 3]
```diff
@@ -271,9 +285,12 @@ def create_xml(recid, texkey):
 
 
 def task_run_core():
-    """ Performs a search to find records without a texkey, generates a new
-    one and uploads the changes in chunks """
-    recids = perform_request_search(p='-035:spirestex -035:inspiretex', cc='HEP')
+    """Perform a search to find records without a texkey.
+
+    generates a new one and uploads the changes in chunks
```
Comment (jirikuncar) [inline invenio/modules/sequencegenerator/texkey.py]: Generates ... chunks.

File: invenio/modules/oauthclient/utils.py [hunk 4]
```diff
@@ -125,6 +125,6 @@ def oauth_link_external_id(user, external_id=None):
 
 def oauth_unlink_external_id(external_id):
     """Unlink a user from an external id."""
-    UserEXT.query.filter_by(id=external_id['id'],
+    UserEXT.query.filter_by(id=external_id['id'].encode('utf-8'),
```
Comment (jirikuncar) [inline invenio/modules/oauthclient/utils.py]: why do you have to encode the value here? it should not be necessary.

**Comment by jacquerie** [comment]
I'll expand on the problem in `inveniomanage demosite populate --yes-i-know`.

`invenio/base/scripts/demosite.py`'s core logic is implemented in these lines: https://github.com/inveniosoftware/invenio/blob/ca72262a23e2660a87614998588b58928f9ab744/invenio/base/scripts/demosite.py#L110-L138. Several legacy applications are called as if they were functions using a wrapper, which also passes them their arguments. The problem is that these tasks could be added in a different order to the `schTASK` queue (implemented as a table) compared to the one `demosite.py` is expecting.

For example, https://github.com/inveniosoftware/invenio/blob/ca72262a23e2660a87614998588b58928f9ab744/invenio/legacy/bibsched/bibtask.py#L1026 tries to find a task with `id = 1` and `proc = bibupload`, but the first row of the `schTASK` table is a `bibsort` task:
```SQL
jnotarst=# select * from "schTASK";
-[ RECORD 1 ]----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
id         | 1
proc       | bibsort
host       | PCGS2X3160H65
user       | admin
runtime    | 2015-06-09 17:35:44
sleeptime  | 
arguments  | \x5b040000007307000000626962736f727473020000002d75730500000061646d696e730d0000002d2d6c6f61642d636f6e666967
status     | DONE
progress   | Will execute: 'bibsort' '-u' 'admin' '--load-config'
priority   | 0
sequenceid |
```

**Comment by jirikuncar** [comment]
@jacquerie @hachreak we need to reset sequence on `schTASK` or drop and create table again.

**Comment by jacquerie** [comment]
So the idea would be to add the task, truncate the table, always query for `id = 1`?

Another idea, suggested by @kaplun: instead of using [`task_init`](https://github.com/inveniosoftware/invenio/blob/ca72262a23e2660a87614998588b58928f9ab744/invenio/legacy/bibsched/bibtask.py#L455), use the pythonic API provided by [`task_low_level_submission`](https://github.com/inveniosoftware/invenio/blob/ca72262a23e2660a87614998588b58928f9ab744/invenio/legacy/bibsched/bibtask.py#L155), which returns the `id` of the enqueued task.

**Comment by jirikuncar** [comment]
> task_low_level_submission

:+1:

**Comment by hachreak** [comment]
ping @jirikuncar :shipit:

File: invenio/modules/access/control.py [hunk 5]
```diff
@@ -1804,11 +1804,11 @@ def acc_delete_all_settings():
     from invenio.ext.sqlalchemy import db
     db.session.commit()
 
-    run_sql("""TRUNCATE "accROLE" """)
-    run_sql("""TRUNCATE "accACTION" """)
-    run_sql("""TRUNCATE "accARGUMENT" """)
-    run_sql("""TRUNCATE "user_accROLE" """)
-    run_sql("""TRUNCATE "accROLE_accACTION_accARGUMENT" """)
+    truncate_table("accROLE")
```
Comment (jirikuncar) [inline invenio/modules/access/control.py]: Can you use the model if you are changing `run_sql` in `invenio.modules.*`? Thanks

**Comment by jirikuncar** [comment]
@hachreak Please remove unnecessary parts from `invenio/modules/access/admin_lib.py` instead of amending them. Thanks

File: invenio/modules/access/control.py (hunk 5 above)
Comment (hachreak) [inline invenio/modules/access/control.py]: e.g. `AccROLE.query.delete()`?

Comment (jirikuncar) [inline invenio/modules/access/control.py]: :+1:

**Pull request merged**